packages/@aws-cdk/aws-appmesh/lib/tls-certificate.ts

```diff
@@ -0,0 +1,199 @@
+import { ICertificate } from '@aws-cdk/aws-certificatemanager';
+import { CfnVirtualGateway, CfnVirtualNode } from './appmesh.generated';
+
+/**
+ * Enum of supported TLS modes
+ */
+export enum TlsMode {
+
+  /**
+   * Only accept encrypted traffic
+   */
+  STRICT = 'STRICT',
+
+  /**
+   * Accept encrypted and plaintext traffic.
+   */
+  PERMISSIVE = 'PERMISSIVE',
+
+  /**
+   * TLS is disabled, only accept plaintext traffic.
+   */
+  DISABLED = 'DISABLED',
+}
+
+/**
+ * Redundant API shapes are represented here.
+ */
+export interface TlsCertificateConfig {
+
+  /**
+   * The CFN shape for a virtual gateway listener TLS certificate
+   */
+  readonly virtualGatewayListenerTlsCertificate: CfnVirtualGateway.VirtualGatewayListenerTlsCertificateProperty,
+
+  /**
+   * The CFN shape for a virtual node listener TLS certificate
+   */
+  readonly virtualNodeListenerTlsCertificate: CfnVirtualNode.ListenerTlsCertificateProperty,
+}
+
+/**
+ * Represents a TLS certificate
+ */
+export abstract class TlsCertificate {
+
+  /**
+   * Returns an File TLS Certificate
+   */
+  public static file(props: FileCertificateOptions): TlsCertificate {
+    return new FileTlsCertificate(props);
+  }
+
+  /**
+   * Returns an ACM TLS Certificate
+   */
+  public static acm(props: ACMCertificateOptions): TlsCertificate {
+    return new ACMTlsCertificate(props);
+  }
+
+  /**
+   * The TLS mode.
+   *
+   * @default - none
+   */
+  readonly abstract tlsMode: TlsMode;
+
+  /**
+   * Returns TLS certificate based provider.
+   */
+  public abstract bind(): TlsCertificateConfig;
+
+}
+
+/**
+ * Represents a ACM provided TLS certificate
+ */
+export class ACMTlsCertificate extends TlsCertificate {
+
+  /**
+   * The TLS mode.
+   *
+   * @default - none
+   */
+  readonly tlsMode: TlsMode;
+
+  /**
+   * The ARN of the ACM certificate
+   */
+  readonly acmCertificate: ICertificate;
+
+  constructor(props: ACMCertificateOptions) {
+    super();
+    this.tlsMode = props.tlsMode;
+    this.acmCertificate = props.acmCertificate;
+  }
+
+  bind(): TlsCertificateConfig {
+    return {
+      virtualGatewayListenerTlsCertificate: {
+        acm: {
+          certificateArn: this.acmCertificate.certificateArn,
+        },
+      },
+      virtualNodeListenerTlsCertificate: {
+        acm: {
+          certificateArn: this.acmCertificate.certificateArn,
+        },
+      },
+    };
+  }
+}
+
+/**
+ * Represents a file provided TLS certificate
+ */
+export class FileTlsCertificate extends TlsCertificate {
+
```

Comment:
I think we should remove this extra line.

Reply:
Dominic wanted an empty line before variables... does that convention not apply to the first?

Reply:
The empty line at the start of class definition can be removed. This is what I observed is being followed in the rest of the files and packages.
for e.g. 
```ts
/**
 * Interface with base properties all routers willl inherit
 */
export interface VirtualRouterBaseProps {
  /**
   * Listener specification for the VirtualRouter
   *
   * @default - A listener on HTTP port 8080
   */
  readonly listeners?: VirtualRouterListener[];

  /**
   * The name of the VirtualRouter
   *
   * @default - A name is automatically determined
   */
  readonly virtualRouterName?: string;
}
```
No empty line at the start of the class and empty line after each variable. 🙂